### PR TITLE
sd-boot: add support for custom mode.

### PR DESCRIPTION
**Pull request by kraxel**
Custom mode allows to write updates to db, dbx, KEK and PK without signature.  See the comment block for a more detailed description.

In case the PK update has no signature try to enable custom mode.

### CODE REVIEW

File: src/boot/efi/secure-boot.c [hunk 1]
```diff
@@ -32,10 +32,46 @@ SecureBootMode secure_boot_mode(void) {
         return decode_secure_boot_mode(secure, audit, deployed, setup);
 }
 
+/*
+ * Custom mode allows to change secure boot certificate databases db, dbx, KEK and PK without the variable
+ * updates being signed.  When enrolling certificates to an unconfigured system (no PK present yet) writing
+ * db, dbx and KEK updates without signature works fine even in standard mode.  Writing PK updates without
+ * signature requires custom mode in any case.
+ *
+ * Enabling custom mode worky only if a user is physically present.  Note that OVMF has a dummy
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: Can you add a link to some spec or document that explains this mode?

Also there are some double spaces in this comment

File: src/boot/efi/secure-boot.c [hunk 2]
```diff
@@ -32,10 +32,46 @@ SecureBootMode secure_boot_mode(void) {
...
+        return enabled;
+}
+
+static EFI_STATUS set_custom_mode(bool enable) {
+        static char16_t name[] = u"CustomMode";
+        static uint32_t attr =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS;
+        uint8_t mode = enable
+                ? 1   // CUSTOM_SECURE_BOOT_MODE
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: `/* */` for comments

File: src/boot/efi/secure-boot.c [hunk 3]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
...
+                        }
+                }
+        }
+
+        if (need_custom_mode && !custom_mode_enabled()) {
+                err = set_custom_mode(true);
+                if (err == EFI_SUCCESS) {
+                        printf("Custom mode enabled.\n");
+                } else {
+                        log_error_status(err, "Enable custom mode: %m");
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: "Failed to enable custom mode:"

File: src/boot/efi/secure-boot.c [hunk 4]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
...
+                        if (*sigsize <= 24) {
+                                printf("PK is not signed (need custom mode).\n");
+                                need_custom_mode = true;
+                        }
+                }
+        }
+
+        if (need_custom_mode && !custom_mode_enabled()) {
+                err = set_custom_mode(true);
+                if (err == EFI_SUCCESS) {
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: no brackets around single line ifs

File: src/boot/efi/secure-boot.c [hunk 5]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
...
+                        uint32_t *sigsize = (uint32_t*)(sb_vars[i].buffer + 16);
+                        if (*sigsize <= 24) {
+                                printf("PK is not signed (need custom mode).\n");
+                                need_custom_mode = true;
+                        }
+                }
+        }
+
+        if (need_custom_mode && !custom_mode_enabled()) {
+                err = set_custom_mode(true);
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: `set_custom_mode(/* enable= */ true)`

File: src/boot/efi/secure-boot.c [hunk 6]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
                         log_error_status(err, "Failed reading file %ls\\%ls: %m", path, sb_vars[i].filename);
                         goto out_deallocate;
                 }
+                if (strcmp(sb_vars[i].name, u"PK") == 0 && sb_vars[i].size > 20) {
```
Comment (bluca) [inline src/boot/efi/secure-boot.c]: streq()

File: src/boot/efi/secure-boot.c [hunk 7]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
...
+                        uint32_t *sigsize = (uint32_t*)(sb_vars[i].buffer + 16);
```
Comment (bluca) [inline src/boot/efi/secure-boot.c:143]: size_t

Comment (bluca) [inline src/boot/efi/secure-boot.c:151]: a few magic numbers in this block, at least a comment explaining what they are would be good

File: src/boot/efi/secure-boot.c [hunk 8]
```diff
@@ -103,6 +139,23 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path, bool
...
+                        if (*sigsize <= 24) {
```
Comment (bluca) [inline src/boot/efi/secure-boot.c:152]: assert(sigsize) for good measure

File: src/boot/efi/secure-boot.c (hunk 1 above)
Comment (kraxel) [reply src/boot/efi/secure-boot.c]: I'm not aware of any good documentation, which is why I wrote that lengthly comment in the first place.
It's not part of the uefi spec, but it is implemented in tianocore and therefore kind-of standard.

File: src/boot/efi/secure-boot.c (hunk 8 above)
Comment (kraxel) [reply src/boot/efi/secure-boot.c:152]: Why?  Doesn't make sense to me.

Comment (bluca) [reply src/boot/efi/secure-boot.c:152]: We have as a code style rule to always assert pointers we get before deref (unless there's an explicit check ofc)

File: src/boot/efi/secure-boot.c (hunk 1 above)
Comment (bluca) [reply src/boot/efi/secure-boot.c]: thanks, looks fine then

File: src/boot/efi/secure-boot.c [hunk 9]
```diff
@@ -32,10 +32,46 @@ SecureBootMode secure_boot_mode(void) {
         return decode_secure_boot_mode(secure, audit, deployed, setup);
 }
 
+/*
+ * Custom mode allows to change secure boot certificate databases db, dbx, KEK and PK without the variable
+ * updates being signed. When enrolling certificates to an unconfigured system (no PK present yet) writing
+ * db, dbx and KEK updates without signature works fine even in standard mode. Writing PK updates without
+ * signature requires custom mode in any case.
+ *
+ * Enabling custom mode worky only if a user is physically present. Note that OVMF has a dummy
```
Comment (yuwata) [inline src/boot/efi/secure-boot.c]: worky -> works ?

File: src/boot/efi/secure-boot.c (hunk 8 above)
Comment (yuwata) [reply src/boot/efi/secure-boot.c:152]: Instead, adding assert for `sb_vars[i].buffer` is better I think.

Comment (kraxel) [reply src/boot/efi/secure-boot.c:152]: Asserting `buffer` does make sense indeed (that should not be NULL if `size > 0`).

**Review by yuwata — APPROVED** [review]

**Review by bluca — APPROVED** [review]
Needs some documentation too, but feel free to open a follow-up pr for that

**Comment by kraxel** [comment]
> Needs some documentation too, but feel free to open a follow-up pr for that

The whole secure boot key enrollment process appears to not be documented very well.  There is a brief paragraph in the systemd-boot man page that apparently expects people already know what those `*.auth` files are and how to create them.  So not sure this is a good place to explain what custom mode is.

Suggestions?

**Comment by bluca** [comment]
True it's underdocumented, then it's fine to leave as is for now

**Pull request merged**